<a href="https://colab.research.google.com/github/yurigalindo/DaemonPS/blob/main/Problem_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Problem 2: Portfolio Rebalance


In [2]:
import warnings
warnings.filterwarnings("ignore") #This ignores all warnings, useful for generating a readable notebook.
#Warnings should not be ignored when developing since they may contain useful info.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Based on the log returns, we can exponentiate the log returns and multiply by the original deltas to obtain the new deltas. We can sum the log returns of the first ten days.

Since the value of the portfolio has changed, the total sum of weights may be different from one. From this total sum we can recover the total return of the portfolio.

We must divide all the deltas by this sum to keep them summing to one before computing the trades needed for rebalancing.

In [4]:
first_deltas = np.array([0.1,0.15,0.1,0.05,0.05,0.1,0.2,0.1,0.05,0.1] )
logreturns = pd.read_csv('logreturns.csv')
logreturns.drop('DATE',1,inplace=True)
#dropped the date column because it doesn't add extra information 
#and math will be smoother without it

In [5]:
ten_days = logreturns.iloc[:10,:] #Select all of the columns for rows 0-9
ten_days_return = ten_days.sum(axis=0) #Computes the sum for each column
ten_days_return = np.exp(ten_days_return.values) #Convert to real returns
ten_days_delta = np.multiply(first_deltas,ten_days_return) #Compute the updated deltas
ten_days_total_return = np.sum(ten_days_delta) #The sum represents the portfolio total value
ten_days_delta /= ten_days_total_return #Make them sum to one
print(ten_days_delta)

[0.09532127 0.17827229 0.07413902 0.05361232 0.05049917 0.10575837
 0.20175461 0.10540594 0.05087038 0.08436663]


In [6]:
ten_days_desired = np.array([0.2,0.1, 0.1, 0.1, 0.05, 0.1, 0.05, 0.2, 0.05, 0.05])
trades = ten_days_desired - ten_days_delta
print(trades)

[ 0.10467873 -0.07827229  0.02586098  0.04638768 -0.00049917 -0.00575837
 -0.15175461  0.09459406 -0.00087038 -0.03436663]


##a) Necessary trades (positive means buy, negative means sell): 

AA    0.104679  
BB   -0.078272  
CC    0.025861  
DD    0.046388  
EE   -0.000499  
FF   -0.005758  
GG   -0.151755  
HH    0.094594  
II   -0.000870  
JJ   -0.034367  

We can repeat the first steps, but this time starting with the new deltas and considering the returns of the final days.

In [8]:
final_days = logreturns.iloc[10:,:] #Select all of the columns for rows 10-14
final_days_return = final_days.sum(axis=0) #Computes the sum for each column
final_days_return = np.exp(final_days_return.values) #Convert to real returns
final_days_delta = np.multiply(ten_days_desired,final_days_return) #Compute the updated deltas
final_days_total_return = np.sum(final_days_delta) #The sum represents the portfolio value
print(ten_days_total_return)
print(final_days_total_return)

1.0182782358231413
0.9915445786832808


Since we consider compound growth, we must multiply the returns of the two periods.

In [10]:
total_return = ten_days_total_return*final_days_total_return
print(total_return)
print((total_return-1)*100)

1.009668264321611
0.9668264321611053


##b) Total return: 0.9%